pkg/api/proxy.go

```diff
@@ -0,0 +1,123 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/dchest/siphash"
+	"github.com/gorilla/mux"
+
+	"zotregistry.dev/zot/pkg/api/config"
+)
+
+func ClusterProxy(ctrlr *Controller) func(http.HandlerFunc) http.HandlerFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			config := ctrlr.Config
+			logger := ctrlr.Log
+
+			// if no cluster or single-node cluster, handle locally
+			if config.Cluster == nil || len(config.Cluster.Members) == 1 {
+				next.ServeHTTP(response, request)
+
+				return
+			}
+
+			vars := mux.Vars(request)
+			name, ok := vars["name"]
+
+			if !ok || name == "" {
+				response.WriteHeader(http.StatusNotFound)
+
+				return
+			}
+
+			h := siphash.New([]byte(config.Cluster.HashKey))
+			h.Write([]byte(name))
+			sum64 := h.Sum64()
+
+			targetMember := config.Cluster.Members[sum64%uint64(len(config.Cluster.Members))]
```

Comment:
Is there a possibility to loose access to content (by being redirected to bad member) if I add or remove cluster members (number of cluster members changes)?

Reply:
Yes, a node can go down and cluster has to be re-shaped.